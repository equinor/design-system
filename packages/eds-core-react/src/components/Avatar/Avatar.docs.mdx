import { Links } from './../../../.storybook/components'
import { Meta, Primary, Canvas, Controls } from '@storybook/blocks'
import * as ComponentStories from './Avatar.stories'

<Meta of={ComponentStories} />

# Avatar

Avatars use images to visually represent people or companies.<br />
The avatar does not exist in Figma as a separate component.

<Links 
    documentationUrl="https://eds.equinor.com/0b0c666ab/p/96f1b1-image-placeholder/b/85d6bc"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Avatar/Avatar.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Primary />
<Controls />

## Usage

<ul>
    <li>Place avatars alongside additional information about the people or companies the avatars represent.</li>
    <li>Avatars can be displayed in lists that are searchable and selectable.</li>
    <li>Use inline avatars to mention people in a sentence.</li>
</ul>

```tsx
import { Avatar } from '@equinor/eds-core-react'

<Avatar />
```

## Examples

### Sizes

<Canvas of={ComponentStories.Size} />

### Disabled

<Canvas of={ComponentStories.Disabled} />

### Context

<Canvas of={ComponentStories.Context} />
