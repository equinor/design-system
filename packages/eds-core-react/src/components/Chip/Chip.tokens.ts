import { tokens } from '@equinor/eds-tokens'
import type { ComponentToken } from '@equinor/eds-tokens'

const {
  spacings: {
    comfortable: { x_small, small, medium },
  },
  typography: {
    ui: { chip__badge: chipTypography },
  },
  colors: {
    ui: {
      background__default: { rgba: backgroundColorDefault },
      background__light: { rgba: backgroundColor },
    },
    interactive: {
      primary__resting: { rgba: primaryColor },
      primary__hover_alt: { rgba: primaryHoverAlt },
      primary__hover: { rgba: primaryHover },
      disabled__text: { rgba: disabledColor },
      focus: { rgba: focusOutlineColor },
      primary__selected_highlight: { rgba: activeColor },
      danger__resting: { rgba: errorColor },
      danger__hover: { rgba: errorColorHover },
      danger__highlight: { rgba: errorBackgroundHover },
    },
  },
  shape: {
    rounded: { borderRadius },
  },
  interactions: {
    focused: { width: focusOutlineWidth },
  },
} = tokens

export type ChipToken = ComponentToken & {
  entities: { icon: ComponentToken }
}

export const enabled: ChipToken = {
  background: backgroundColor,
  height: '22px',
  border: {
    radius: borderRadius,
    color: 'transparent',
    type: 'border',
    width: '1px',
    style: 'solid',
  },
  spacings: {
    left: x_small,
    right: small,
  },
  typography: {
    ...chipTypography,
    color: primaryColor,
  },

  states: {
    hover: {
      typography: {
        color: primaryHover,
      },
    },
    disabled: {
      typography: {
        color: disabledColor,
      },
    },
    focus: {
      outline: {
        width: focusOutlineWidth,
        color: focusOutlineColor,
        style: 'dashed',
        type: 'outline',
        offset: '3px',
      },
    },
    active: {
      background: activeColor,
    },
  },
  entities: {
    icon: {
      height: medium,
      width: medium,
      border: {
        radius: borderRadius,
        type: 'border',
        width: 0,
      },
      states: {
        hover: {
          background: primaryHoverAlt,
        },
      },
    },
  },
}

export const error: ChipToken = {
  background: backgroundColorDefault,
  border: {
    type: 'border',
    color: errorColor,
    width: '1px',
    style: 'solid',
    radius: borderRadius,
  },
  typography: {
    color: errorColor,
  },
  states: {
    hover: {
      typography: {
        color: errorColorHover,
      },
    },
  },
  entities: {
    icon: {
      typography: {
        color: errorColor,
      },
      background: errorBackgroundHover,
    },
  },
}
