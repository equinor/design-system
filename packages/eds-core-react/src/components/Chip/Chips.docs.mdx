import { Links, PropsTable, Story } from './../../../.storybook/components'
import { Chip } from '.'

# Chip

Chips, also known as tags or badges, are advanced badges that represent discrete information.

<Links 
    figmaUrl="https://www.figma.com/file/0bGXR2sCwMVSDNyyzu5BXrO5/UI%E2%80%94User-Interface?node-id=17497%3A0"
    documentationUrl="https://eds.equinor.com/0b0c666ab/p/95e67f-chip/b/95ad56"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Chip/Chip.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Story id="data-display-chips--introduction"/>
<PropsTable />

## Usage

Chips allow users to enter information, make selections, filter content, or trigger actions.
<ul>
    <li>Chips are <strong>compact</strong> components that represent discrete information.</li>
    <li>Chips should be <strong>relevant</strong> to the content or task they represent.</li>
    <li>Chips should make tasks easier to complete, or content easier to sort.</li>
</ul>

  ```tsx
import { Chip } from '@equinor/eds-core-react'

<Chip>tag</Chip>
```

## Examples

### Text

<Story id="data-display-chips--text" />

### Text and Icon

<Story id="data-display-chips--text-and-icon" />

### Text and Avatar

<Story id="data-display-chips--text-and-avatar" />
