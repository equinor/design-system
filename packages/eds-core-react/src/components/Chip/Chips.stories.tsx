import { action } from '@storybook/addon-actions'
import styled from 'styled-components'
import { Icon, Chip, ChipProps, Avatar, AvatarProps } from '../..'
import { ComponentMeta, Story } from '@storybook/react'
import { save } from '@equinor/eds-icons'
import page from './Chips.docs.mdx'

const icons = {
  save,
}

Icon.add(icons)

const Wrapper = styled.div`
  display: grid;
  grid-gap: 32px;
  grid-template-columns: repeat(4, fit-content(100%));
`

export default {
  title: 'Data Display/Chips',
  component: Chip,
  parameters: {
    docs: {
      page,
    },
  },
} as ComponentMeta<typeof Chip>

const handleDelete = action('onDelete')
const handleClick = action('onClick')

const CatImage = (props: Partial<AvatarProps>) => (
  <Avatar src={'https://i.imgur.com/UM3mrju.jpg'} alt="cat" {...props} />
)

export const Introduction: Story<ChipProps> = (args) => (
  <Wrapper>
    <Chip {...args}>Play with me</Chip>
  </Wrapper>
)

export const Text: Story<ChipProps> = () => (
  <Wrapper>
    <Chip>Normal</Chip>
    <Chip variant="active">Active</Chip>
    <Chip variant="error">Error</Chip>
    <Chip disabled>Disabled</Chip>
    <Chip onClick={handleClick}>clickable</Chip>
    <Chip variant="active" onClick={handleClick}>
      clickable
    </Chip>
    <Chip variant="error" onClick={handleClick}>
      clickable
    </Chip>
    <Chip onDelete={handleDelete} disabled>
      deletable
    </Chip>
    <Chip onDelete={handleDelete}>deletable</Chip>
    <Chip variant="active" onDelete={handleDelete}>
      deletable
    </Chip>
    <Chip variant="error" onDelete={handleDelete}>
      deletable
    </Chip>
    <div></div>
    <Chip onDelete={handleDelete} onClick={handleClick}>
      deletable + clickable
    </Chip>
    <Chip variant="active" onDelete={handleDelete} onClick={handleClick}>
      deletable + clickable
    </Chip>
    <Chip variant="error" onDelete={handleDelete} onClick={handleClick}>
      deletable + clickable
    </Chip>
  </Wrapper>
)

export const TextAndIcon: Story<ChipProps> = () => (
  <Wrapper>
    <Chip>
      <Icon name="save" />
      Normal
    </Chip>
    <Chip variant="active">
      <Icon name="save" />
      Active
    </Chip>
    <Chip variant="error">
      <Icon name="save" />
      Error
    </Chip>
    <Chip disabled>
      <Icon name="save" />
      Disabled
    </Chip>
    <Chip onClick={handleClick}>
      <Icon name="save" />
      clickable
    </Chip>
    <Chip variant="active" onClick={handleClick}>
      <Icon name="save" />
      clickable
    </Chip>
    <Chip variant="error" onClick={handleClick}>
      <Icon name="save" />
      clickable
    </Chip>
    <Chip onDelete={handleDelete} disabled>
      <Icon name="save" />
      deletable
    </Chip>
    <Chip onDelete={handleDelete}>
      <Icon name="save" />
      deletable
    </Chip>
    <Chip variant="active" onDelete={handleDelete}>
      <Icon name="save" />
      deletable
    </Chip>
    <Chip variant="error" onDelete={handleDelete}>
      <Icon name="save" />
      deletable
    </Chip>
    <div></div>
    <Chip onDelete={handleDelete} onClick={handleClick}>
      <Icon name="save" />
      deletable + clickable
    </Chip>
    <Chip variant="active" onDelete={handleDelete} onClick={handleClick}>
      <Icon name="save" />
      deletable + clickable
    </Chip>
    <Chip variant="error" onDelete={handleDelete} onClick={handleClick}>
      <Icon name="save" />
      deletable + clickable
    </Chip>
  </Wrapper>
)
export const TextAndAvatar: Story<ChipProps> = () => (
  <Wrapper>
    <Chip>
      <CatImage />
      Normal
    </Chip>
    <Chip variant="active">
      <CatImage />
      Active
    </Chip>
    <Chip variant="error">
      <CatImage />
      Error
    </Chip>
    <Chip disabled>
      <CatImage />
      Disabled
    </Chip>
    <Chip onClick={handleClick}>
      <CatImage />
      clickable
    </Chip>
    <Chip variant="active" onClick={handleClick}>
      <CatImage />
      clickable
    </Chip>
    <Chip variant="error" onClick={handleClick}>
      <CatImage />
      clickable
    </Chip>
    <Chip onDelete={handleDelete} disabled>
      <CatImage />
      deletable
    </Chip>
    <Chip onDelete={handleDelete}>
      <CatImage />
      deletable
    </Chip>
    <Chip variant="active" onDelete={handleDelete}>
      <CatImage />
      deletable
    </Chip>
    <Chip variant="error" onDelete={handleDelete}>
      <CatImage />
      deletable
    </Chip>
    <div></div>
    <Chip onDelete={handleDelete} onClick={handleClick}>
      <CatImage />
      deletable + clickable
    </Chip>
    <Chip variant="active" onDelete={handleDelete} onClick={handleClick}>
      <CatImage />
      deletable + clickable
    </Chip>
    <Chip variant="error" onDelete={handleDelete} onClick={handleClick}>
      <CatImage />
      deletable + clickable
    </Chip>
  </Wrapper>
)

TextAndIcon.storyName = 'Text and icon'
TextAndAvatar.storyName = 'Text and avatar'
