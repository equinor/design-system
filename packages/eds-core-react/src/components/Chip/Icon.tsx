import styled, { css } from 'styled-components'
import { close } from '@equinor/eds-icons'
import { Icon as Icon_ } from '../Icon'
import * as tokens from './Chip.tokens'
import { bordersTemplate } from '@equinor/eds-utils'

Icon_.add({ close })

const { enabled, error } = tokens

type IconProps = {
  $variant: 'active' | 'error' | 'default'
  $disabled: boolean
}

export const Icon = styled(Icon_)<IconProps>`
  cursor: pointer;
  padding: 1px;
  ${bordersTemplate(enabled.entities.icon.border)}
  &:hover {
    ${({ $variant }) => {
      switch ($variant) {
        case 'error':
          return css`
            background: ${error.entities.icon.background};
            svg {
              fill: ${error.entities.icon.typography.color};
            }
          `
        default:
          return css`
            background: ${enabled.entities.icon.states.hover.background};
            svg {
              fill: ${enabled.states.hover.typography.color};
            }
          `
      }
    }};
  }

  ${({ $disabled }) =>
    $disabled &&
    css`
      cursor: not-allowed;
      &:hover {
        background: transparent;
      }
    `}
`
