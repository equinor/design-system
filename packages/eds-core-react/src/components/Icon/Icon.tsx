import { forwardRef, Ref, SVGProps } from 'react'
import type { IconData } from '@equinor/eds-icons'
import styled from 'styled-components'
import type { Name } from './Icon.types'
import { get } from './library'

type StyledProps = {
  height: number
  $height?: number
  width: number
  fill: string
  size?: number
  rotation?: number
}

type SimpleSVGProps = {
  name: string
  viewBox: string
  rotation?: number
  title?: string
  role?: string
  'aria-hidden'?: boolean | 'true' | 'false'
  'aria-labelledby'?: string
}

const transform = ({ rotation }: SimpleSVGProps): string =>
  rotation ? `transform: rotate(${rotation}deg)` : ''

const StyledSvg = styled.svg.attrs<StyledProps>(({ height, width, fill }) => ({
  name: null,
  xmlns: 'http://www.w3.org/2000/svg',
  height: `${height}px`,
  width: `${width}px`,
  fill,
}))`
  ${transform}
`

const StyledPath = styled.path.attrs<StyledProps>(({ $height, size }) => ({
  size: null,
  fillRule: 'evenodd',
  clipRule: 'evenodd',
  transform: size / $height !== 1 ? `scale(${size / $height})` : null,
}))``

const customIcon = (icon: IconData) => ({
  icon,
  count: Math.floor(Math.random() * 1000),
})

const findIcon = (name: string, data: IconData, size: number) => {
  // eslint-disable-next-line prefer-const
  let { icon, count } = data ? customIcon(data) : get(name)

  if (size < 24) {
    // fallback to normal icon if small is not made yet
    icon = icon.sizes?.small || icon
  }

  return { icon, count }
}

export type IconProps = {
  /** Title for icon when used semantically */
  title?: string
  /** Color */
  color?: string
  /** Size */
  size?: 16 | 18 | 24 | 32 | 40 | 48
  /** Rotation */
  rotation?: 0 | 90 | 180 | 270
  /** Name */
  name?: Name
  /** Manually specify which icon data to use */
  data?: IconData
  /** @ignore */
  ref?: Ref<SVGSVGElement>
} & SVGProps<SVGSVGElement>

export const Icon = forwardRef<SVGSVGElement, IconProps>(function Icon(
  { size, color = 'currentColor', name, rotation, title, data, ...rest },
  ref,
) {
  // eslint-disable-next-line prefer-const
  const { icon, count } = findIcon(name, data, size)

  if (typeof icon === 'undefined') {
    throw Error(
      `Icon "${name}" not found. Have you added it using Icon.add() or using data props?`,
    )
  }

  const height = size ? size : parseInt(icon.width)
  const width = size ? size : parseInt(icon.height)

  let svgProps: SimpleSVGProps & StyledProps = {
    height,
    width,
    fill: color,
    viewBox: `0 0 ${width} ${height}`,
    rotation,
    name,
    'aria-hidden': true,
  }

  const pathProps = {
    d: icon.svgPathData,
    $height: icon.height ? icon.height : size,
    size: size || icon.height,
  }

  // Accessibility
  let titleId = ''

  if (title) {
    titleId = `${icon.prefix}-${icon.name}-${count}`
    svgProps = {
      ...svgProps,
      title,
      role: 'img',
      'aria-hidden': undefined,
      'aria-labelledby': titleId,
    }
  }

  return (
    <StyledSvg {...svgProps} {...rest} ref={ref}>
      {title && <title id={titleId}>{title}</title>}
      <StyledPath data-testid="eds-icon-path" {...pathProps} />
    </StyledSvg>
  )
})

// Icon.displayName = 'EdsIcon'
