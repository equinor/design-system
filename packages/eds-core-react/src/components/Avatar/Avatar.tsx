import { forwardRef, HTMLAttributes } from 'react'
import styled, { css } from 'styled-components'
import { bordersTemplate } from '@equinor/eds-utils'
import { avatar as tokens } from './Avatar.tokens'

type StyledAvatarProps = {
  $size: number
}

const StyledAvatar = styled.div<StyledAvatarProps>`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  overflow: hidden;
  ${bordersTemplate(tokens.border)};
  ${({ $size }) => css`
    height: ${$size}px;
    width: ${$size}px;
  `}
`

type StyledImageProps = {
  alt: string
  src: string
  $disabled: boolean
}

const StyledImage = styled.img<StyledImageProps>`
  height: 100%;
  text-align: center;
  color: transparent;

  ${({ $disabled }) =>
    $disabled &&
    css`
      opacity: ${tokens.states.disabled.opacity};
    `};
`

export type AvatarProps = {
  alt: string
  /** Image source
   @default null */
  src?: string
  /** Avatar size
  @default 24 */
  size?: 16 | 24 | 32 | 40 | 48
  /** @default false */
  disabled?: boolean
} & HTMLAttributes<HTMLDivElement>

export const Avatar = forwardRef<HTMLDivElement, AvatarProps>(function Avatar(
  { src = null, alt, size = 24, disabled = false, ...rest },
  ref,
) {
  return (
    <StyledAvatar $size={size} ref={ref} {...rest}>
      <StyledImage src={src} alt={alt} $disabled={disabled} />
    </StyledAvatar>
  )
})

Avatar.displayName = 'Avatar'
