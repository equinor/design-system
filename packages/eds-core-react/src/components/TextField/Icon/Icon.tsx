import { HTMLAttributes, forwardRef } from 'react'
import styled, { css } from 'styled-components'
import { useTextField } from '../TextField.context'
import { input as tokens } from './Icon.tokens'
import type { Variants, ColorStateProps } from '../types'

type StyledIconProps = {
  colors: ColorStateProps
  isDisabled?: boolean
  isFocused?: boolean
  size?: 16 | 24
}

const StyledIcon = styled.div<StyledIconProps>(
  ({ colors, isDisabled, isFocused, size }) => {
    const { focusColor, color, disabledColor } = colors
    let fill = color

    if (isDisabled) {
      fill = disabledColor
    }
    if (isFocused) {
      fill = focusColor
    }

    return css`
      &,
      svg {
        fill: ${fill};
        width: ${size}px;
        height: ${size}px;
      }
    `
  },
)

type TextfieldIconProps = {
  /** isDisabled */
  isDisabled?: boolean
  /** Variant */
  variant?: Variants
  /** Colors */
  colors?: ColorStateProps
  /** Size */
  size?: 16 | 24
} & HTMLAttributes<HTMLDivElement>

const InputIcon = forwardRef<HTMLDivElement, TextfieldIconProps>(
  function InputIcon(
    {
      size = 24,
      variant = 'default',
      isDisabled = false,
      colors = {
        color: tokens[variant].color,
        disabledColor: tokens[variant].disabledColor,
        focusColor: tokens[variant].focusColor,
      },
      children,
      ...other
    },
    ref,
  ) {
    const { isFocused } = useTextField()

    const iconProps = {
      isDisabled,
      colors,
      isFocused,
      size,
    }

    return (
      <StyledIcon ref={ref} {...iconProps} {...other}>
        {children}
      </StyledIcon>
    )
  },
)

export { InputIcon as Icon }
