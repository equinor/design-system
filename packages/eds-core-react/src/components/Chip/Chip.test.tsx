import { render, fireEvent, screen } from '@testing-library/react'
import { axe } from 'jest-axe'
import styled from 'styled-components'
import { add } from '@equinor/eds-icons'
import { Chip } from './Chip'
import { Avatar } from '../Avatar'
import { Icon } from '../Icon'
import * as tokens from './Chip.tokens'

const StyledChips = styled(Chip)`
  position: relative;
`

const { enabled, error } = tokens

describe('Chips', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(
      <Chip>
        <Icon data={add} />
        Chip
      </Chip>,
    )
    expect(asFragment()).toMatchSnapshot()

    const { asFragment: asFragment2 } = render(
      <Chip
        onClick={() => {
          jest.fn()
        }}
      >
        <Icon data={add} />
        Chip
      </Chip>,
    )
    expect(asFragment2()).toMatchSnapshot()
  })
  it('Can extend the css for the component', () => {
    render(<StyledChips>styled chip</StyledChips>)
    expect(screen.getByText('styled chip')).toHaveStyleRule(
      'position',
      'relative',
    )
  })
  it('Should pass a11y test', async () => {
    const { container } = render(<Chip>I am chip</Chip>)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test when has icon', async () => {
    const { container } = render(
      <Chip>
        <Icon data={add} />I am chip
      </Chip>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test when has avatar', async () => {
    const imageUrl = 'https://i.imgur.com/UM3mrju.jpg'
    const { container } = render(
      <Chip>
        <Avatar src={imageUrl} alt="avatar" />I am chip
      </Chip>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Has provided text', () => {
    const chipText = 'hello, I am a chip'
    render(<Chip>{chipText}</Chip>)
    expect(screen.getByText(chipText)).toBeDefined()
  })
  it('Has provided Icon', () => {
    const chipText = 'hello, I am a chip'
    const iconTestId = 'avatar-chip-test'
    render(
      <Chip>
        <Icon data={add} data-testid={iconTestId} />
        {chipText}
      </Chip>,
    )
    expect(screen.getByText(chipText)).toBeDefined()
    expect(screen.getByTestId(iconTestId)).toBeDefined()
  })
  it('Has correct styling when Icon is provided', () => {
    const chipText = 'hello, I am a chip'
    const iconTestId = 'avatar-chip-test'
    render(
      <Chip>
        <Icon data={add} data-testid={iconTestId} />
        {chipText}
      </Chip>,
    )
    const chip = screen.queryByText(chipText)
    const icon = screen.queryByTestId(iconTestId)

    expect(chip).toBeDefined()
    expect(chip).toHaveStyleRule('padding-left', '4px')
    expect(chip).toHaveStyleRule('padding-right', enabled.spacings.right)
    expect(chip).toHaveStyleRule(
      'border-radius',
      enabled.border.type === 'border' && enabled.border.radius,
    )
    expect(icon).toBeDefined()
    expect(icon).toHaveAttribute('height', enabled.entities.icon.height)
    expect(icon).toHaveAttribute('width', enabled.entities.icon.width)
  })
  it('Has provided Avatar', () => {
    const chipText = 'hello, I am a chip'
    const avatarTestId = 'avatar-chip-test'
    const imageUrl = 'https://i.imgur.com/UM3mrju.jpg'
    render(
      <Chip>
        <Avatar src={imageUrl} data-testid={avatarTestId} alt="avatar" />
        {chipText}
      </Chip>,
    )
    expect(screen.getByText(chipText)).toBeDefined()
    expect(screen.getByTestId(avatarTestId)).toBeDefined()
    expect(screen.getByAltText('avatar')).toHaveAttribute('src', imageUrl)
  })

  it('Has correct styling when Avatar is provided', () => {
    const chipText = 'hello, I am a chip'
    const avatarTestId = 'avatar-chip-test'
    const imageUrl = 'https://i.imgur.com/UM3mrju.jpg'
    render(
      <Chip>
        <Avatar src={imageUrl} data-testid={avatarTestId} alt="avatar" />
        {chipText}
      </Chip>,
    )
    const chip = screen.getByText(chipText)
    const avatar = screen.queryByTestId(avatarTestId)
    expect(chip).toBeDefined()
    expect(chip).toHaveStyleRule('padding-left', '4px')
    expect(chip).toHaveStyleRule('padding-right', enabled.spacings.right)
    expect(chip).toHaveStyleRule(
      'border-radius',
      enabled.border.type === 'border' && enabled.border.radius,
    )
    expect(avatar).toBeDefined()
    expect(screen.getByAltText('avatar')).toHaveAttribute('src', imageUrl)
    expect(avatar).toHaveStyleRule('height', enabled.entities.icon.height)
    expect(avatar).toHaveStyleRule('width', enabled.entities.icon.width)
  })

  it('Has called handleDelete once with props when close icon is clicked', () => {
    const chipText = 'hello, I am a chip'
    const chipId = 'avatar-chip-test'
    let callbackId = ''
    const handleDelete = jest.fn((x: React.MouseEvent<HTMLElement>) => {
      // eslint-disable-next-line testing-library/no-node-access
      callbackId = (x.target as HTMLElement).parentElement.id
    })

    render(
      <Chip id={chipId} onDelete={handleDelete}>
        {chipText}
      </Chip>,
    )

    const closeIcon = screen.queryAllByTitle('close')[0]

    fireEvent.click(closeIcon)

    expect(handleDelete).toHaveBeenCalled()
    expect(callbackId).toEqual(chipId)
  })
  it('Has called handleClick once with props when clicked', () => {
    const chipText = 'hello, I am a chip'
    const chipId = 'avatar-chip-test'
    let callbackId = ''
    const handleClick = jest.fn((x: React.MouseEvent<HTMLElement>) => {
      callbackId = (x.target as HTMLElement).id
    })

    render(
      <Chip id={chipId} onClick={handleClick} data-testid="chip">
        {chipText}
      </Chip>,
    )

    fireEvent.click(screen.getByTestId('chip'))

    expect(handleClick).toHaveBeenCalled()
    expect(callbackId).toEqual(chipId)
  })
  it('Has called handleClick once with props when pressed Enter', () => {
    const chipText = 'hello, I am a chip'
    const chipId = 'avatar-chip-test'
    let callbackId = ''
    const handleClick = jest.fn((x: React.MouseEvent<HTMLElement>) => {
      callbackId = (x.target as HTMLElement).id
    })

    render(
      <Chip id={chipId} onClick={handleClick} data-testid="chip">
        {chipText}
      </Chip>,
    )

    fireEvent.click(screen.getByTestId('chip'), {
      key: 'Enter',
    })

    expect(handleClick).toHaveBeenCalled()
    expect(callbackId).toEqual(chipId)
  })
  it('Has correct active styling', () => {
    const chipText = 'hello, I am a chip'
    render(<Chip variant="active">{chipText}</Chip>)
    expect(screen.queryByText(chipText)).toHaveStyleRule(
      'background',
      enabled.states.active.background,
    )
  })
  it('Has some correct error styling', () => {
    const chipText = 'hello, I am a chip'
    const iconTestId = 'avatar-chip-test'

    render(
      <Chip variant="error">
        <Icon data={add} data-testid={iconTestId} />
        {chipText}
      </Chip>,
    )
    const chip = screen.queryByText(chipText)
    expect(chip).toHaveStyleRule('background', error.background)
    expect(chip).toHaveStyleRule('color', error.typography.color)
  })
})
