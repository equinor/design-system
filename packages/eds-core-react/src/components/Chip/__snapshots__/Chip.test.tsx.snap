// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chips Matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  background: var(--eds_ui_background__light,rgba(247,247,247,1));
  height: 22px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  display: grid;
  grid-gap: 8px;
  grid-auto-flow: column;
  grid-auto-columns: max-content;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid transparent;
  border-radius: 100px;
  padding-left: 4px;
  padding-right: 8px;
  margin: 0;
  color: var(--eds_interactive_primary__resting,rgba(0,112,121,1));
  font-family: Equinor;
  font-size: 0.750rem;
  font-weight: 500;
  line-height: 1.333em;
  text-align: left;
}

.c0 svg {
  fill: var(--eds_interactive_primary__resting,rgba(0,112,121,1));
}

.c0:focus {
  outline: none;
}

.c0[data-focus-visible-added]:focus {
  outline: 1px dashed var(--eds_interactive__focus,rgba(0,112,121,1));
  outline-offset: 2px;
}

.c0:focus-visible {
  outline: 1px dashed var(--eds_interactive__focus,rgba(0,112,121,1));
  outline-offset: 2px;
}

@media (hover:hover) and (pointer:fine) {
  .c0:hover {
    color: var(--eds_interactive_primary__hover,rgba(0,79,85,1));
  }

  .c0:hover svg {
    fill: var(--eds_interactive_primary__hover,rgba(0,79,85,1));
  }
}

<div
    class="c0"
  >
    <svg
      aria-hidden="true"
      class=""
      fill="currentColor"
      height="16px"
      viewBox="0 0 16 16"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        clip-rule="evenodd"
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        data-testid="eds-icon-path"
        fill-rule="evenodd"
        height="24"
        transform="scale(0.6666666666666666)"
      />
    </svg>
    Chip
  </div>
</DocumentFragment>
`;
